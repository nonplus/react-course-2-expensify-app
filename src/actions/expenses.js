import uuid from "uuid";
import database from "../firebase/firebase";
import _ from "lodash";

export const addExpense = expense => ({
  type: "ADD_EXPENSE",
  expense
});

export const addExpenseAsync = (expenseData = {}) => {
  return async dispatch => {
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };
    const ref = await database.ref("expenses").push(expense);
    await dispatch(addExpense({ id: ref.key, ...expense }));
  };
};

export const setExpenses = expenses => ({
  type: "SET_EXPENSES",
  expenses
});

export const setExpenseAsync = () => {
  return async dispatch => {
    const snapshot = await database.ref("expenses").once("value");
    const expenses = _.map(snapshot.val(), (value, id) => ({ id, ...value }));
    dispatch(setExpenses(expenses));
  };
};

export const editExpense = (id, updates) => {
  const { description, note, amount, createdAt } = updates;
  return {
    type: "EDIT_EXPENSE",
    id,
    updates: {
      description,
      note,
      amount,
      createdAt
    }
  };
};

export const editExpenseAsync = (id, updates) => {
  const { description, note, amount, createdAt } = updates;
  return async dispatch => {
    await database
      .ref(`expenses/${id}`)
      .set({ description, note, amount, createdAt });
    dispatch(editExpense(id, updates));
  };
};

export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id
});

export const removeExpenseAsync = ({ id } = {}) => {
  return async dispatch => {
    await database.ref(`expenses/${id}`).remove();
    dispatch(removeExpense({ id }));
  };
};
