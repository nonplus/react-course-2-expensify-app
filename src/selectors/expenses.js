export const selectExpenses = (
  expenses,
  { text, sortBy, startDate, endDate }
) => {
  const startDateMatches = startDate
    ? expense => startDate.isSameOrBefore(expense.createdAt, "day")
    : () => true;
  const endDateMatches = endDate
    ? expense => endDate.isSameOrAfter(expense.createdAt, "day")
    : () => true;
  text = text.toLowerCase();
  const textMatches = !text
    ? () => true
    : expense => expense.description.toLowerCase().includes(text);

  const sortCompare =
    sortBy === "date"
      ? (a, b) => b.createdAt - a.createdAt
      : sortBy === "amount" ? (a, b) => b.amount - a.amount : () => 0;

  return expenses
    .filter(
      expense =>
        startDateMatches(expense) &&
        endDateMatches(expense) &&
        textMatches(expense)
    )
    .sort(sortCompare);
};
