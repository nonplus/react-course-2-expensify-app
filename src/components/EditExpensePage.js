import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { editExpense, removeExpense } from "../actions/expenses";

export const EditExpensePage = ({
  editExpense,
  removeExpense,
  history,
  expense
}) => {
  const onSubmit = changedExpense => {
    editExpense(expense.id, changedExpense);
    history.push("/");
  };

  const onRemove = () => {
    removeExpense({ id: expense.id });
    history.push("/");
  };

  return (
    <div>
      <h1>Edit Expense</h1>
      <ExpenseForm expense={expense} onSubmit={onSubmit} />
      <button onClick={onRemove}>Remove</button>
    </div>
  );
};

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

const mapDispatchToProps = dispatch => ({
  editExpense: (id, changedExpense) =>
    dispatch(editExpense(id, changedExpense)),
  removeExpense: expense => dispatch(removeExpense(expense))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
