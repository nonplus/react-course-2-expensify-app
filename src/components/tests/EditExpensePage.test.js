import React from "react";
import { shallow } from "enzyme";
import expenses from "../../tests/fixtures/expenses";
import { EditExpensePage } from "../EditExpensePage";

describe("<EditExpensePage />", () => {
  let editExpenseAsync, removeExpenseAsync, history, wrapper, expense;

  beforeEach(() => {
    expense = { ...expenses[1] };
    editExpenseAsync = jest.fn();
    removeExpenseAsync = jest.fn();
    history = {
      push: jest.fn()
    };
    wrapper = shallow(
      <EditExpensePage
        expense={expense}
        editExpenseAsync={editExpenseAsync}
        removeExpenseAsync={removeExpenseAsync}
        history={history}
      />
    );
  });

  it("should render the page correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe("ExpenseForm.onSubmit", () => {
    let editExpenseAsyncPromise;
    beforeEach(() => {
      editExpenseAsyncPromise = Promise.resolve(undefined);
      editExpenseAsync.mockReturnValue(editExpenseAsyncPromise);
    });
    it("should call props.editExpenseAsync(expense)", () => {
      const updates = { ...expense };
      delete updates.id;
      wrapper.find("ExpenseForm").prop("onSubmit")(updates);
      expect(editExpenseAsync).toHaveBeenLastCalledWith(expense.id, updates);
    });
    it("should call history.push('/')", async () => {
      wrapper.find("ExpenseForm").prop("onSubmit")(expenses[1]);
      await editExpenseAsyncPromise;
      expect(history.push).toHaveBeenLastCalledWith("/");
    });
  });

  describe("button.onClick", () => {
    const removeExpenseAsyncPromise = Promise.resolve(undefined);
    beforeEach(() => {
      removeExpenseAsync.mockReturnValue(removeExpenseAsyncPromise);
    });
    it("should call props.removeExpense(expense)", () => {
      wrapper.find("button").prop("onClick")();
      expect(removeExpenseAsync).toHaveBeenLastCalledWith({ id: expense.id });
    });
    it("should call history.push('/')", async () => {
      wrapper.find("button").prop("onClick")();
      await removeExpenseAsyncPromise;
      expect(history.push).toHaveBeenLastCalledWith("/");
    });
  });
});
