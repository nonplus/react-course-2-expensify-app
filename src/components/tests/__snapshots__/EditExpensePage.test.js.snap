// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditExpensePage /> should render the page correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 109500,
          "createdAt": -345600000,
          "description": "Rent",
          "id": "id_2",
          "note": "",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      Remove Expense
    </button>
  </div>
</div>
`;
